llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -11133,6 +11133,18 @@ SDValue DAGCombiner::visitCTLZ_ZERO_UNDEF(SDNode *N) {
   if (SDValue C =
           DAG.FoldConstantArithmetic(ISD::CTLZ_ZERO_UNDEF, DL, VT, {N0}))
     return C;
+
+  // Fold ctlz_zero_undef(X << C) --> ctlz_zero_undef(X) - C
```

Comment:
It's not obvious to me that it's better. it's just trading a shift for an add or sub?

Reply:
Is it better to fold `C1 - ctlz_zero_undef(X << C2) --> C1 + C2 - ctlz_zero_undef(X)`?

Reply:
You can do both. The win is probably that the parameter of the `ctlz_zero_undef` got simpler.